[Boolector] Add support for const arrays and boolean indices/elements

nbailluet opened the pull request:
The conversion of "const array" terms was not yet implemented for Boolector.
This PR implements the conversion of pysmt array types into btor array sorts:
https://github.com/pysmt/pysmt/blob/743ff01f994b57437db6251866dbc39074f494de/pysmt/solvers/btor.py#L532-L534

and add support for building `ConstArray` instances with associated array stores (inpired from both msat and z3 converters):
https://github.com/pysmt/pysmt/blob/743ff01f994b57437db6251866dbc39074f494de/pysmt/solvers/btor.py#L513-L521

It also removes the `test_btor_does_not_support_const_arrays` and replace it with a test to ensure nothing breaks when building a const array with Boolector:
https://github.com/pysmt/pysmt/blob/743ff01f994b57437db6251866dbc39074f494de/pysmt/test/test_array.py#L79-L83

mikand commented:
Hi @nbailluet, thanks a lot for this!

One thing that is missing is to declare in the `LOGICS` of the btor solver that the array value is now supported (see: https://github.com/pysmt/pysmt/blob/f6e2c66975077da842d3cb8c58a8c7a8b500fb42/pysmt/logics.py#L722C1-L722C20)

For the rest, LGTM!

mikand commented:
@nbailluet Another thing now that the tests have been executed: I know it is tedious, but please add the email (160679793+nbailluet@users.noreply.github.com) you used for the git commit in the `.mailmap` file to signify that you agree with the pysmt DCO. Thanks! 

See: https://pysmt.readthedocs.io/en/latest/development.html

nbailluet commented:
> @nbailluet Another thing now that the tests have been executed: I know it is tedious, but please add the email ([160679793+nbailluet@users.noreply.github.com](mailto:160679793+nbailluet@users.noreply.github.com)) you used for the git commit in the `.mailmap` file to signify that you agree with the pysmt DCO. Thanks!
> 
> See: https://pysmt.readthedocs.io/en/latest/development.html

Actually I have commited with the email I've already added to `CONTRIBUTORS`. However my email was kept private on my github profile, I've changed the setting, so now I guess it'll work as intended.

mikand commented:
Not sure this is enough, because in git the masked mail would remain...
Try to run the `ci/cla.sh` script locally, to see if it works. Otherwise you can either add the masked email to the `.mailmap` file or rewrite the git history and force push with the new email

nbailluet commented:
> Hi @nbailluet, thanks a lot for this!
> 
> One thing that is missing is to declare in the `LOGICS` of the btor solver that the array value is now supported (see: https://github.com/pysmt/pysmt/blob/f6e2c66975077da842d3cb8c58a8c7a8b500fb42/pysmt/logics.py#L722C1-L722C20)
> 
> For the rest, LGTM!

@mikand The latest commit add the relevant logics to the `LOGICS` field by filtering the `ARRAYS_CONST_LOGIC` on the logics' name, not sure how to do that otherwise.

nbailluet commented:
> Not sure this is enough, because in git the masked mail would remain... Try to run the `ci/cla.sh` script locally, to see if it works. Otherwise you can either add the masked email to the `.mailmap` file or rewrite the git history and force push with the new email

@mikand It does work locally, but I'm gonna add it to the mailmap I guess.

mikand commented:
@nbailluet Apart for the CLA issue, the code now looks perfect!

A.

nbailluet commented:
> @nbailluet Apart for the CLA issue, the code now looks perfect!

@mikand I've re-commited with the good email, now it should work, sorry for that.

nbailluet commented:
@mikand I've implemented back-conversion of const arrays in `get_value` and added support for boolean indices and elements by using bit-vectors of size 1.
I've also added some missing tests for Boolector and an example of QF_ABV* formula in `get_example_formulae`.
Would you mind taking a look at these new commits and tell me if it looks good to you?

mikand reviewed (approved):
LGTM!

mikand commented:
@nbailluet Thanks a lot for this, it looks great!
